### DISCUSSION TITLE
Help: TypeError: list indices must be integers or slices, not str in Flow but not in Code Component

### DISCUSSION THREAD

**Discussion by FBR65** [langgenius/dify#8834] [Help]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I'm calling the Knowledge and getting this Answer:

```
{
  "result": [
    {
      "metadata": {
        "_source": "knowledge",
        "position": 1,
        "dataset_id": "30d3abd8-eeac-49c4-95ab-5cb0e07bd9e9",
        "dataset_name": "Knowledge Base",
        "document_id": "a1b04523-e802-4e7c-b448-0a8864e8edcf",
        "document_name": "1234-5678-9123",
        "document_data_source_type": "upload_file",
        "segment_id": "c57cfeec-b5af-4d9a-819a-d6588e77b7a7",
        "retriever_from": "workflow",
        "score": 0.95606345,
        "segment_hit_count": 5,
        "segment_word_count": 2144,
        "segment_position": 10,
        "segment_index_node_hash": "e1334738c6caf32fdd0ebcceb4f0cce3cfd028ce21f8219b0328b8dd8f980060"
      },
      "title": "1234-5678-9123",
      "content": "XXXXXXX"
    }
  ]
}
```

In the next Step I'm using the Code Component with this Code:

```
def main(data: str, user: str) -> str:
    first_result = data["result"][0]
    if first_result["metadata"]["position"] == 1 and first_result["metadata"]["document_name"] == user:
        return {"result": "OK"}
    else:
        return {"result": "Nope"}
```

Testing the Code inside the Code Component with user "emil" works fine.

Result is as expected ("result":"Nope" ). No Error is thrown.

Inside the Flow i'm always getting the Error:

```
Run failed: Traceback (most recent call last):
File "/var/sandbox/sandbox-python/tmp/274210ea_af06_444c_93c9_866199b2fc72.py", line 48, in
File "", line 20, in
File "", line 4, in main
TypeError: list indices must be integers or slices, not str
error: exit status 255
```

I can't figure out why this happens.

### 2. Additional context or comments

I need this kind of functionality to ensure that only the user who uploaded the knowledge can evaluate it. This is basically a personal RAG.

**Comment by FBR65** [comment]
Hi, it was a Layer 8 problem.  I didn't get that the answer from th knowledge isn't a dictionary but a list.

Correct Code:

import json

def main(data_list: list, str, user: str) -> str:
    data_dict = json.loads(data_list)
    first_result = data_dict["result"][0]["metadata"]["document_name"]
    if str(first_result) == user:
        return {"result": "OK"}
    else:
        return {"result": "Nope"}

**Reply by Lishan-Jin** [reply]
I'm now facing exactly the same problem. Would u please provide more detail about how u fix it? The code above seems can't compose.
